### PR TITLE
[SPARK-10382] Make example code in user guide testable

### PR DESCRIPTION
**Pull request by yinxusen**
A POC code for making example code in user guide testable.

@mengxr We still need to talk about the labels in code.

### CODE REVIEW

File: docs/_config.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@ highlighter: pygments
 markdown: kramdown
 gems:
   - jekyll-redirect-from
+  - octopress-code-highlighter
```
Comment (mengxr) [inline docs/_config.yml]: Do we still need this gem if we have `include_example.rb`?

File: docs/_plugins/include_example.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
+require 'octopress-code-highlighter'
+require 'liquid'
+
+module Octopress
```
Comment (mengxr) [inline docs/_plugins/include_example.rb]: If our implementation doesn't require octopress, we should rename it and cite it in the reference.

File: docs/_plugins/include_example.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        site = context.registers[:site]
+        config_dir = (site.config['code_dir'] || '../examples/src/main').sub(/^\//,'')
+        @code_dir = File.join(site.source, config_dir)
+
+        begin
+          options = get_options
+          code = File.open(@file).read.encode("UTF-8")
+          code = select_lines(code, options)
+
+          CodeHighlighter.highlight(code, options)
```
Comment (mengxr) [inline docs/_plugins/include_example.rb]: Can we generate `{% highlight scala %}` directly? We don't need anything fancy here.

File: docs/_plugins/include_example.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        options[:lang] ||= File.extname(@file).delete('.')
+        options[:link_text] ||= "Example code"
+        options
+      end
+
+      # Select lines according to labels in code. Currently we use "begin code" and "end code" as
+      # labels.
+      def select_lines(code, options)
+        lines = code.each_line.to_a
+        start = lines.each_with_index.select { |l, i| l.include? "begin code" }.last.last
```
Comment (mengxr) [inline docs/_plugins/include_example.rb]: "begin code" and "end code" are quite common. Shall we switch to `$example on$` and `$example off$`?

File: docs/index.md [hunk 5]
```diff
@@ -12,6 +12,8 @@ It also supports a rich set of higher-level tools including [Spark SQL](sql-prog
 
 # Downloading
 
+{% include_example scala/org/apache/spark/examples/SparkALS.scala SparkALS %}
```
Comment (mengxr) [inline docs/index.md]: Could turn it into a real example including Scala/Java/Python? We can use the pipeline example http://spark.apache.org/docs/latest/ml-guide.html#example-pipeline.

File: docs/_plugins/include_example.rb (hunk 3 above)
Comment (yinxusen) [inline docs/_plugins/include_example.rb]: So there is no need to call outside function, just translate it into markdown highlight. Then we can avoid to use Octopress in the beginning. I get it.

File: docs/_plugins/include_example.rb (hunk 4 above)
Comment (yinxusen) [inline docs/_plugins/include_example.rb]: Here is the thing, different language requires different comment signs. So in Java and Scala we have `// example on` and `// example off` (or `/* example on/off */`) and in Python, it is `# example on/off`. Does your `$example on/off$` mean that we use the dollar sign to distinguish our label with common code comments?

File: docs/index.md (hunk 5 above)
Comment (yinxusen) [inline docs/index.md]: Sure I will add it after refactoring the code.

File: docs/_plugins/include_example.rb (hunk 4 above)
Comment (mengxr) [inline docs/_plugins/include_example.rb]: We don't check the comment symbol like "//" or "#". Just do a string match on the content. So both of the following work. I use `$` just to make sure the text doesn't appear in normal comments.

~~~scala
// $example on$
~~~

~~~python
# $example off$
~~~

**Comment by yinxusen** [comment]
@mengxr I add a POC example of ml#Pipeline. But I am not sure that I select right files. But it demos what we need.

**Comment by yinxusen** [comment]
It requires Pygments. So install Pygments before compile it. Actually it is the same with what defined in _config.yml.

**Comment by mengxr** [comment]
@yinxusen It is possible that we only replace the markdown text and let `highlight` handle the syntax highlighting? Another minor comment is removing the `// $example on$` and `// $example off$` in the output file. Could you also include the imports in the example code? Basically, having two example blocks from the source file. Just want to make sure it works.

**Comment by mengxr** [comment]
@yinxusen Could you update the PR to support multiple example code blocks in one file? It is quite comment to have imports + example code - boilerplate code in the user guide. If you show that it works, we can remove the example and then merge this PR.

**Comment by yinxusen** [comment]
@mengxr Yes, I'll do it ASAP.

**Comment by yinxusen** [comment]
@mengxr You can check the POC now. If it is OK, I will remove the example. Next time we can enjoy easy code sample insertion in docs.

File: docs/_plugins/include_example.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
+require 'liquid'
```
Comment (mengxr) [inline docs/_plugins/include_example.rb]: Could you add Apache header? See `copy_api_dirs.rb`.

**Comment by mengxr** [comment]
Yes, it looks good. Please remove the example, add Apache header, then remove `[POC]` from the PR title. It should be good to go.

File: docs/_plugins/include_example.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ 
+      Pygments.highlight(code, :lexer => @lang)
+    end
+ 
+    # Trim the code block so as to have the same indention, regardless of their positions in the
+    # code file.
+    def trim_codeblock(lines)
+      # Select the minimum indention of the current code block.
+      min_start_spaces = lines
+        .select{ |l| l.strip.size !=0 }
```
Comment (mengxr) [inline docs/_plugins/include_example.rb]: nit: if we still want to follow Spark code style, it should be `select {` instead of `select{`

Comment (yinxusen) [inline docs/_plugins/include_example.rb]: Sure, will fix it.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks! This makes the example code much easier to check. Could you make one JIRA and submit a PR to replace some example code using this? Then we can create more JIRAs, and ask community to help.

**Comment by yinxusen** [comment]
@mengxr Sure I can do that.

**Pull request closed without merging**